File: python/amostraPequena/casoAleatorio.py
import random
import time

def heapify(arr, n, i):
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    if left < n and arr[left] > arr[largest]:
        largest = left
    if right < n and arr[right] > arr[largest]:
        largest = right
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)

def heap_sort(arr):
    n = len(arr)
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)
    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heapify(arr, i, 0)
arr = list(range(1, 10001))
random.shuffle(arr)

print("Realizando Heap Sort com 10.000 elementos (caso aleatorio)")
inicio = time.time()

heap_sort(arr)

fim = time.time()
print(f"\nTempo total: {fim - inicio:.4f} segundos")
print("Está ordenado corretamente?", arr == sorted(arr))

File: C/amostraPequena/casoAleatorio.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void heapify(int arr[], int n, int i) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest])
        largest = left;
    
    if (right < n && arr[right] > arr[largest])
        largest = right;
    
    if (largest != i) {
        int temp = arr[i];
        arr[i] = arr[largest];
        arr[largest] = temp;
        heapify(arr, n, largest);
    }
}

void heapSort(int arr[], int n) {
    for (int i = n / 2 - 1; i >= 0; i--)
        heapify(arr, n, i);

    for (int i = n - 1; i > 0; i--) {
        int temp = arr[0];
        arr[0] = arr[i];
        arr[i] = temp;
        heapify(arr, i, 0);
    }
}

int main() {
    int n = 10000;
    int *arr = (int *)malloc(n * sizeof(int));
    if (!arr) {
        printf("Falha ao alocar memória.\n");
        return 1;
    }

    for (int i = 0; i < n; i++)
        arr[i] = i + 1;

    for (int i = n - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    printf("Iniciando ordenação com Heap Sort de caso aleatorio de 10.000 elementos...\n");
    clock_t start = clock();

    heapSort(arr, n);

    clock_t end = clock();
    double time_spent = (double)(end - start) / CLOCKS_PER_SEC;
    printf("Tempo total: %.4f segundos\n", time_spent);

    free(arr);
    return 0;
}